### PR TITLE
Move Tekton onto the injection-gen based controller infrastructure

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

Migrates the controller code to [`knative/pkg` injection approach](https://github.com/knative/pkg/tree/master/injection) as done in [`knative/build`](https://github.com/knative/build/pull/617) by @mattmoor

> […] the goal of the broader refactoring was to effectively eliminate boilerplate outside of `pkg/reconciler/foo` for writing controllers, and keep our various controllers from drifting apart as we add features (e.g. this gives Build various metrics integrations including stackdriver that I believe it lacked previously).
> 
> This also enables us to pretty trivially consolidate controllers, or split them apart by simply changing what goes into the main binary.

Marking this as wip because : 
- [x] I need to broke it down into several commits (to ease review)
- [ ] Fix the one test in `pipelinerun` that check stuff with the logs (and I'm not yet sure how to do that)
- [ ] *Clean the mess*, aka remove variables, fields that are not required anymore, simplify some signatures

/cc @ImJasonH @bobcatfish @mattmoor

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ :no_good_man: ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

# Release Notes

Not sure if it requires a release not item but…

```
Switching to  [`knative/pkg` injection approach](https://github.com/knative/pkg/tree/master/injection) for the pipeline controller
```

### CODE REVIEW

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go [hunk 1]
```diff
@@ -362,6 +360,8 @@ func TestReconcile_InvalidPipelineRuns(t *testing.T) {
 		})
 	}
 }
+
+/*
 func TestReconcile_InvalidPipelineRunNames(t *testing.T) {
```
Comment (vdemeester) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go]: This is the only test that I can't figure yet how to "migrate" as it uses the `logs` from testAssets that aren't available (as a result of the controller change). I wonder if there is a better way to test this without having to "parse" the logs

**Comment by bobcatfish** [comment]
Hey @vdemeester  what (if anything) are the consequences of not including `defer cancel()` in the reconciler level tests?

**Comment by vdemeester** [comment]
@bobcatfish the linter won't be happy as it is required to call `cancel`, so if we don't, we may have some goroutine lying around. On the tests it's not a huge deal but still :stuck_out_tongue_closed_eyes: 

>  Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete.

**Pull request merged**